### PR TITLE
Big MNIST model with PIL augmentation and load/save 

### PR DESCRIPTION
**Pull request by marcelbischoff**
`test_mnist2` needs to be merged into `serious_mnist`

### CODE REVIEW

**Comment by geohot** [comment]
So this is a lot to merge. Let's do just batch norm first and get those tests passing.

File: test/test_nn.py [hunk 1]
```diff
@@ -8,33 +8,40 @@ class TestNN(unittest.TestCase):
...
-    np.testing.assert_allclose(outt.data, toutt.detach().numpy(), rtol=1e-5)
+    for training in [False, True]:
+      # create in tinygrad
+      bn = BatchNorm2D(sz, eps=1e-5, training=training)
+      bn.weight = Tensor.randn(sz)
+      bn.bias = Tensor.randn(sz)
+      bn.running_mean = Tensor.randn(sz)
+      bn.running_var = Tensor.randn(sz)
+      bn.running_var.data[bn.running_var.data < 0] = 0
+      #debug
```
Comment (geohot) [inline test/test_nn.py:19]: Remove?

File: test/test_nn.py [hunk 2]
```diff
@@ -8,33 +8,40 @@ class TestNN(unittest.TestCase):
...
+      bn.running_mean = Tensor.zeros(sz)
+      bn.running_var = Tensor.zeros(sz)
+
+      # create in torch
+      #with torch.no_grad():
+      if True:
```
Comment (geohot) [inline test/test_nn.py:25]: if True?

File: examples/train_efficientnet.py [hunk 3]
```diff
@@ -37,19 +37,21 @@ def load_cifar():
 
   Tensor.default_gpu = os.getenv("GPU") is not None
   TINY = os.getenv("TINY") is not None
+  TRANSFER = os.getenv("TRANSFER") is not None
```
Comment (geohot) [inline examples/train_efficientnet.py:40]: I'd merge just this file right now.

**Comment by geohot** [comment]
Got your BatchNorm merged

https://github.com/geohot/tinygrad/commit/ffb96b2d0b5f34a0acb63af0312871fd532138f4

**Comment by marcelbischoff** [comment]
```loss 0.01 accuracy 1.00:  98%|██████████████████████████████████████████████████▊ | 3906/4000 [8:29:45<12:33,  8.01s/it]
```
hahaha

**Comment by kartik4949** [comment]
Testing acc?

**Comment by marcelbischoff** [comment]
Yes, current batch test accuracy which is not surprising. 

ToDo:
* Final dense layer has no bias!

**Comment by marcelbischoff** [comment]
after 12h 🥁 

```
tinygrad % ipython test/test_mnist2.py          
(128, 1, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 10)
(128, 10)
(1, 10)
no of parameters 1187606
loss 0.00 accuracy 1.00: 100%|████████████████████████████████████████████████████| 4000/4000 [8:42:17<00:00,  7.83s/it]
(128, 1, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 10)
(128, 10)
(1, 10)
no of parameters 1187606
loss 0.00 accuracy 1.00: 100%|████████████████████████████████████████████████████| 1000/1000 [2:14:13<00:00,  8.05s/it]
(128, 1, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128, 128, 3, 3)
(128,)
(128,)
(128, 4)
(1, 4)
(4, 128)
(1, 128)
(128, 10)
(128, 10)
(1, 10)
no of parameters 1187606
loss 0.00 accuracy 1.00: 100%|████████████████████████████████████████████████████| 1000/1000 [2:18:23<00:00,  8.30s/it]
100%|███████████████████████████████████████████████████████████████████████████████████| 78/78 [02:20<00:00,  1.80s/it]
test set accuracy is 0.991700
```

**Comment by geohot** [comment]
Oh man, we need load and save!

**Comment by marcelbischoff** [comment]
Yes, I am regretting I didn't do this before I start the run. I am just tidying this up on a new local branch. I got load and save but need to test.

**Comment by marcelbischoff** [comment]
Tidied up a lot. Still having a separate file `2serious_mnist.py` and just checking if it still learns. On GPU it is impossible slow....

File: examples/2serious_mnist.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    return se
+
+class ConvBlock:
+  def __init__(self, h, w, inp, filters=128, conv=3):
+    self.h = h
+    self.w = w
+    self.filters = filters
+    self.conv = conv
+    self.inp = inp
+    self.c1 = Tensor.uniform(filters, inp, conv, conv)
```
Comment (geohot) [inline examples/2serious_mnist.py:41]: Biases on the convs?

File: test/test_mnist.py [hunk 5]
```diff
@@ -85,10 +85,12 @@ def train(model, optim, steps, BS=128, gpu=False):
     accuracies.append(accuracy)
     t.set_description("loss %.2f accuracy %.2f" % (loss, accuracy))
 
-def evaluate(model, gpu=False):
+def evaluate(model, gpu=False, BS=128):
```
Comment (geohot) [inline test/test_mnist.py:88]: Let's move these functions (train/evaluate) to extra/training.py

File: examples/2serious_mnist.py (hunk 4 above)
Comment (marcelbischoff) [reply examples/2serious_mnist.py:41]: I definitely missed that ones.... keras hides so many things! That's why this is a good exercise!

**Comment by marcelbischoff** [comment]
weird 

```
loss -0.04 accuracy 0.99:  12%|██▉                     | 492/4000 [1:06:17<8:02:17,  8.25s/it]
```

**Comment by marcelbischoff** [comment]
Doing another test run without biases. GPU is still slow as hell (I checked that I am not using OpenCL CPU).

**Comment by marcelbischoff** [comment]
`test set accuracy is 0.992800`

**Comment by kartik4949** [comment]
@marcelbischoff can u try with label smoothning and tta for slight improvement!

**Review by marcelbischoff — COMMENTED** [review]
> @marcelbischoff can u try with label smoothning and tta for slight improvement!

So far there is zero augmentation. I actually briefly tried TTA on the TF code

**Comment by marcelbischoff** [comment]
adding padding but it gets 3x slower

**Comment by marcelbischoff** [comment]
It seems to be working well know.

**Comment by marcelbischoff** [comment]
Augmentation using PIL aug

**Comment by geohot** [comment]
Delete my serious_mnist.py and remove the checkpoints from the repo and I'll merge
